contracts/cw3-flex-multisig/src/state.rs

```diff
@@ -45,6 +86,57 @@ impl Proposal {
 
         status
     }
+
+    /// update_status sets the status of the proposal to current_status.
+    /// (designed for handler logic)
+    pub fn update_status(&mut self, block: &BlockInfo) {
+        self.status = self.current_status(block);
+    }
+
+    // returns true iff this proposal is sure to pass (even before expiration if no future
+    // sequence of possible votes can cause it to fail)
+    pub fn is_passed(&self, block: &BlockInfo) -> bool {
+        match self.threshold {
+            Threshold::AbsoluteCount {
+                weight: weight_needed,
+            } => self.votes.yes >= weight_needed,
+            Threshold::AbsolutePercentage {
+                percentage: percentage_needed,
+            } => self.votes.yes >= votes_needed(self.total_weight, percentage_needed),
+            Threshold::ThresholdQuora { threshold, quorum } => {
+                // this one is tricky, as we have two compares:
+                if self.expires.is_expired(block) {
+                    // * if we have closed yet, we need quorum% of total votes to have voted (counting abstain)
+                    //   and threshold% of yes votes from those who voted (ignoring abstain)
+                    let total = self.votes.total();
+                    let opinions = total - self.votes.abstain;
+                    total >= votes_needed(self.total_weight, quorum)
+                        && self.votes.yes >= votes_needed(opinions, threshold)
+                } else {
+                    // * if we have not closed yet, we need threshold% of yes votes (from 100% voters - abstain)
+                    //   as we are sure this cannot change with any possible sequence of future votes
+                    // * we also need quorum (which may not always be the case above)
+                    self.votes.total() >= votes_needed(self.total_weight, quorum)
+                        && self.votes.yes
+                            >= votes_needed(self.total_weight - self.votes.abstain, threshold)
+                }
+            }
+        }
+    }
+}
+
+// this is a helper function so Decimal works with u64 rather than Uint128
+// also, we must *round up* here, as we need 8, not 7 votes to reach 50% of 15 total
+fn votes_needed(weight: u64, percentage: Decimal) -> u64 {
+    // we multiply by 1million to detect rounding issues
+    const FACTOR: u128 = 1_000_000;
```

Comment:
Better moved to a global (config) section? Better naming would be `PRECISION_FACTOR` or so.

I don't want to introduce unnecessary complexity, but this can be probably transformed into an `u32` (or lower) `PRECISION_DIGITS`, and used as a power of 10 exponent.

Reply:
> Better moved to a global (config) section? Better naming would be PRECISION_FACTOR or so.

Agree with pull it out to a global

> I don't want to introduce unnecessary complexity, but this can be probably transformed into an u32 (or lower) PRECISION_DIGITS, and used as a power of 10 exponent.

It could, but why add an unneeded `.pow()` everytime we calculate a division. Maybe we can do that one time at init

Reply:
Maybe it can be done at init, or even at global scope, and with keeping the `const`.

I'll try it if you don't mind.

Reply:
done